src/lib/CLValue2/Result.ts

```diff
@@ -0,0 +1,85 @@
+import { concat } from '@ethersproject/bytes';
+import { Ok, Err, Result } from 'ts-results';
+
+import { CLType, CLValue, ToBytes } from './Abstract';
+import { toBytesU8 } from '../ByteConverters';
+
+export enum CLErrorCodes {
+  EarlyEndOfStream = 0,
+  Formatting,
+  LeftOverBytes,
+  OutOfMemory
+}
+
+export class CLErrorType extends CLType {
+  toString(): string {
+    return 'Error';
+  }
+}
+
+const RESULT_TAG_ERROR = 1;
+const RESULT_TAG_OK = 1;
+
+/**
+ * Class representing a result of an operation that might have failed. Can contain either a value
+ * resulting from a successful completion of a calculation, or an error. Similar to `Result` in Rust
+ * or `Either` in Haskell.
+ */
+export class CLResultType extends CLType {
+  toString(): string {
+    return 'Result';
+  }
+}
+
+export class CLResult<T extends CLValue & ToBytes, E extends CLErrorCodes>
+  extends CLValue
+  implements ToBytes {
+  public data: T | null;
```

Comment:
Same as with `Option`. Why not to use `Result<T, E>` everywhere to store inner value and be constructor arg?
```
// When constructing, the CLResult should be used.
let r = CLResult.Ok(123);

// Then can be converted to Result.
let v = r.value();

// Now I'd modify it, but its not hand to put it back.
```

I'd rather do:

```
let r = new CLResult(Ok("asd"));
let v = r.value();
let r2 = new CLResult(v);
```